In [5]:
import numpy as np

# Dataset XOR
X = np.array([[0,0],[0,1],[1,0],[1,1]])
y = np.array([[0],[1],[1],[0]])

# Parameter
input_size = 2
hidden_size = 3  # sesuai permintaan
output_size = 1
lr = 0.1

# Inisialisasi bobot (optional: He init untuk ReLU)
W1 = np.random.randn(input_size, hidden_size) * np.sqrt(2 / input_size)
b1 = np.zeros((1, hidden_size))
W2 = np.random.randn(hidden_size, output_size)
b2 = np.zeros((1, output_size))

# Fungsi aktivasi
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def sigmoid_derivative(output):
    return output * (1 - output)

def relu(x):
    return np.maximum(0, x)

def relu_derivative(x):
    return np.where(x > 0, 1, 0)

# Training
for epoch in range(10000):
    # Forward pass
    z1 = np.dot(X, W1) + b1
    a1 = relu(z1)  # gunakan ReLU di hidden layer

    z2 = np.dot(a1, W2) + b2
    a2 = sigmoid(z2)  # output tetap sigmoid karena binary

    # Hitung error
    error = y - a2

    # Backpropagation
    d_a2 = error * sigmoid_derivative(a2)

    d_W2 = np.dot(a1.T, d_a2)
    d_b2 = np.sum(d_a2, axis=0, keepdims=True)

    d_hidden = np.dot(d_a2, W2.T) * relu_derivative(z1)

    d_W1 = np.dot(X.T, d_hidden)
    d_b1 = np.sum(d_hidden, axis=0, keepdims=True)

    # Update bobot
    W1 += lr * d_W1
    b1 += lr * d_b1
    W2 += lr * d_W2
    b2 += lr * d_b2

    # Tampilkan loss setiap 1000 epoch
    if epoch % 1000 == 0:
        loss = np.mean(np.square(error))
        print(f"Epoch {epoch}, Loss: {loss:.6f}")

# Output akhir
print("\nPrediksi Akhir:")
print(a2)


Epoch 0, Loss: 0.219303
Epoch 1000, Loss: 0.005575
Epoch 2000, Loss: 0.001850
Epoch 3000, Loss: 0.001064
Epoch 4000, Loss: 0.000736
Epoch 5000, Loss: 0.000558
Epoch 6000, Loss: 0.000448
Epoch 7000, Loss: 0.000373
Epoch 8000, Loss: 0.000319
Epoch 9000, Loss: 0.000279

Prediksi Akhir:
[[0.01984123]
 [0.9899907 ]
 [0.99000628]
 [0.01984115]]


| Epoch | Loss (Hidden=2) | Loss (Hidden=3) |
|-------|-----------------|-----------------|
| 0 | ~0.25 | **0.2588** |
| 1000 | ~0.22–0.26 | **0.2357** |
| 2000 | ~0.18–0.22 | **0.1803** |
| 3000 | Turun perlahan | **0.0546** |
| 4000 | ~0.05–0.07 | **0.0159** |
| 5000–9000 | Turun bertahap | **Menurun stabil hingga < 0.003** |
| Final Loss | ~0.01 – 0.05 | **0.0025** |

| Epoch | **Dengan ReLU** | **Tanpa ReLU** |
|-------|----------------|----------------|
| 0 | 0.219303 | 0.258775 |
| 1000 | **0.005575** | 0.235703 |
| 2000 | **0.001850** | 0.180301 |
| 3000 | **0.001064** | 0.054600 |
| 4000 | **0.000736** | 0.015945 |
| 5000 | **0.000558** | 0.008273 |
| 6000 | **0.000448** | 0.005411 |
| 7000 | **0.000373** | 0.003968 |
| 8000 | **0.000319** | 0.003111 |
| 9000 | **0.000279** | 0.002548 |
| **Final Loss** | **0.000279** | **0.002548** |

### Perbandingan Prediksi Akhir

| Input | Target | Prediksi ReLU | Prediksi Tanpa ReLU |
|-------|--------|---------------|---------------------|
| [0, 0] | 0 | **0.01984** | 0.02833 |
| [0, 1] | 1 | **0.98999** | 0.94965 |
| [1, 0] | 1 | **0.99001** | 0.95397 |
| [1, 1] | 0 | **0.01984** | 0.05618 |
